Animatable template & animation transparency customization settings

Starkku opened the pull request:
Split from #1300 

-------------

### Customizable animation transparency settings

- `Translucency.Cloaked` can be used to override `Translucency` on animations attached to currently cloaked TechnoTypes.
- Both `Translucency` and `Translucency.Cloaked` can use the new keyframe system to animate along with the animation. Read more about the keyframe system below.
- If interpolation is enabled, the keyframe values are clamped to valid transparency values (0,25,50 and 75), e.g a value of 1.5 would become 0 and 56.525 would become 50 and so on.

In `artmd.ini`:
```ini
[SOMEANIM]             ; AnimationType
Translucency=0         ; integer - only accepted values are 75, 50, 25 and 0.
Translucency.Cloaked=  ; integer - only accepted values are 75, 50, 25 and 0.
```

----------------

### Keyframe animations (Animatable template)

- Some features use keyframe-based animation system to define animations in INI. Defined in INI it looks something like following.

```ini
[SOMESECTION]
BASEKEY.KeyframeN.Value=        ; Key-dependant value type
BASEKEY.KeyframeN.Percentage=   ; floating point value, percents or absolute
BASEKEY.KeyframeN.Absolute=     ; integer, zero-based frame index
BASEKEY.Interpolation=none      ; Interpolation mode (none|linear)
BASEKEY.ResetData=false         ; boolean
```

- `BASEKEY` is whatever base key name the feature in question may use. `N` is zero-based keyframe index. If no keyframes are defined, a single value from `BASEKEY` is attempted to be parsed instead.
  - `Value` is a key/feature-dependant value type associated with that keyframe.
  - `Percentage` is the percentage through the animation's frames where the keyframe becomes active. It is also possible to instead use zero-based frame index via `Absolute` which takes precedence over percentage, albeit it is internally converted to a percentage value.
  - `Interpolation` controls interpolation of values between keyframes. The behaviour here may depend on the value type in use, as not all value types may be interpolatable well or at all.
  - `ResetData` if set to true makes it so that all existing keyframe data is reset before parsing. Can be used to reset keyframes when redefining them in map files etc.

```{note}
Keyframes are expected to be defined with no duplicates for Percentage or Absolute. Failure to do so will crash the game and output developer warnings about offending keys to the log.
```

Starkku committed:
Add animation transparency customization settings

Starkku committed:
Add animation transparency customization settings

Metadorius committed:
Animatable template shenanigans + some template improvements p.1

Metadorius committed:
Animatable template shenanigans + some template improvements p.1

Starkku committed:
Fix percent overriding frame setting

Starkku committed:
Fix percent overriding frame setting

Metadorius committed:
Template is now percentage based, added TExtraArgs... and anim length

the voices

Metadorius committed:
Template is now percentage based, added TExtraArgs... and anim length

the voices

Starkku committed:
Fix and use Animatable template

Starkku committed:
Fix and use Animatable template

Starkku commented:
Now uses @Metadorius's `Animatable` template that allows defining arbitrary keyframe animations in INI.

Might require testing to see if there's a noticeable performance overhead from the template.

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1400,8 +1424,70 @@ if(_strcmpi(parser.value(), #name) == 0){ value = __uuidof(name ## LocomotionCla
+
+	template <>
+	inline double interpolate<double>(double& first, double& second, double percentage, InterpolationMode mode)
+	{
+		double result = first;
+
+		switch (mode)
+		{
+		case InterpolationMode::Linear:
+			result = first + ((second - first) * percentage);
+			break;
+		default:
+			break;
+		}
+
+		return result;
+	}
+
+	template <>
+	inline ColorStruct interpolate<ColorStruct>(ColorStruct& first, ColorStruct& second, double percentage, InterpolationMode mode)
+	{
+		ColorStruct result = first;
+
+		switch (mode)
+		{
+		case InterpolationMode::Linear:
+		{
+			BYTE r = static_cast<BYTE>(first.R + ((second.R - first.R) * percentage));
+			BYTE g = static_cast<BYTE>(first.G + ((second.G - first.G) * percentage));
+			BYTE b = static_cast<BYTE>(first.B + ((second.B - first.B) * percentage));
+			result = ColorStruct { r, g, b };
+			break;
+		}
+		default:
+			break;
+		}
+
+		return result;
+	}
+}
```
I think you can collapse all that by defining a primitive interpolation or two, say, on double, and routing all other specializations to that. Although I would probably make interpolation mode a template param (not sure on that?) and write different functions per each interpolation, and have a single function that dispatches stuff. Could also make it a separate file for clarity, this file is getting quite big.

Starkku replied:
Honestly I question the need for interpolation modes beyond none and linear in the first place anyway.

Metadorius replied:
It depends on the use case. Although I can agree that the specific use cases I am thinking of are rather niche.

What I had in mind was a buildup anim/deploy techno attachment FLH translation, for instance, or maybe just movable FLH/TurretOffset for animated units.

Metadorius replied:
It is probably still good to lay out the ground for easier extensibility while we're at it.

Starkku replied:
> I think you can collapse all that by defining a primitive interpolation or two, say, on double, and routing all other specializations to that

This is heavily case-specific. It works for numeric values like integers (can just truncate the double result) but pretty much anything else does not work like that. ColorStruct, one of the examples I added in wouldn't work like this for an example - you can't just create the color data out of thin air from a floating point value.

Metadorius commented on src/Ext/Anim/Hooks.cpp:
```diff
@@ -329,6 +329,84 @@ DEFINE_HOOK(0x4232BF, AnimClass_DrawIt_MakeInfantry, 0x6)
+
+		if (pTypeExt->Translucent_Keyframes.KeyframeData.size() > 0)
+		{
+			flags |= pTypeExt->Translucent_Keyframes.Get(static_cast<double>(currentFrame) / frames);
```
Isn't there a timer somewhere that will happily already give you a percentage?

Starkku replied:
No. `StageClass` deals with integer increments.

Metadorius commented on src/Ext/Anim/Hooks.cpp:
```diff
@@ -329,6 +329,84 @@ DEFINE_HOOK(0x4232BF, AnimClass_DrawIt_MakeInfantry, 0x6)
+		}
+		else
+		{
+			if (currentFrame > frames * 0.6)
+				flags |= BlitterFlags::TransLucent75;
+			else if (currentFrame > frames * 0.4)
+				flags |= BlitterFlags::TransLucent50;
+			else if (currentFrame > frames * 0.2)
+				flags |= BlitterFlags::TransLucent25;
```
Not sure if we should hardcode that?

Starkku replied:
This is the default `Translucent` behaviour used if the keyframes are not in place.

Metadorius replied:
oh, a comment explaining that would be good in this case

Starkku commented on src/Ext/Anim/Hooks.cpp:
```diff
@@ -329,6 +329,84 @@ DEFINE_HOOK(0x4232BF, AnimClass_DrawIt_MakeInfantry, 0x6)
+				if (translucencyLevel > 15)
+					return ReturnFromFunction;
+				else if (translucencyLevel > 10)
+					flags |= BlitterFlags::TransLucent50;
```
No, not 75.

Starkku commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -470,6 +470,24 @@ In `artmd.ini`:
+  - Keyframes are defined via `Translucent.KeyframeN.*` settings where N is zero-based keyframe index.
+  - `Translucent.KeyframeN.Value` is keyframe's transparency value.
+  - `Translucent.KeyframeN.Percentage` is the percentage through the animation's frames where the keyframes becomes active. It is also possible to instead use zero-based frame index via `Translucent.KeyframeN.Absolute` which takes precedence over percentage.
+  - Keyframes should be listed in ascending order with no duplicate percentages / frames across different keyframes. Failure to do so can produce unpredictable results.
```
Sure if you can think of a way to check for this without adding another massive layer of complexity.

Starkku commented on src/Ext/Anim/Hooks.cpp:
```diff
@@ -329,6 +329,84 @@ DEFINE_HOOK(0x4232BF, AnimClass_DrawIt_MakeInfantry, 0x6)
+			}
+		}
+
+		if (translucency <= 0)
+		{
+			if (translucencyLevel)
+			{
+				if (translucencyLevel > 15)
+					return ReturnFromFunction;
+				else if (translucencyLevel > 10)
+					flags |= BlitterFlags::TransLucent50;
+				else if (translucencyLevel > 5)
+					flags |= BlitterFlags::TransLucent50;
+				else
+					flags |= BlitterFlags::TransLucent25;
+			}
+		}
+		else
+		{
+			if (translucencyLevel >= 15)
+				return ReturnFromFunction;
+			else if (translucency == 75)
+				flags |= BlitterFlags::TransLucent75;
+			else if (translucency == 50)
+				flags |= BlitterFlags::TransLucent50;
+			else if (translucency == 25)
+				flags |= BlitterFlags::TransLucent25;
+		}
+	}
+	else
+	{
+		if (translucencyLevel >= 15)
+			return ReturnFromFunction;
```
Reverse-engineered game code. `translucencyLevel` is some sort of stage/timer thing set on building anims when they need to be drawn at certain transparency when they are normally not, contrary to the YRpp description seems to go from 0 to 15 and is directly copied from the parent building's field `0x6ED`. Rest is just mapping the `Translucency` integer to correct blitter flags.

Perhaps, the translucency stages are basically self-explanatory though and this is probably the only place where whatever the shit `translucencyLevel` is supposed to mean is used.

Metadorius replied:
could leave a comment where stuff is reversed and where custom additions start and we're gucci I htink?

Starkku committed:
Add comments for the unintitiated

Starkku committed:
Add comments for the unintitiated

Starkku committed:
Move interpolation mode enum and functions to separate file

Starkku committed:
Move interpolation mode enum and functions to separate file

Starkku committed:
Fix some small issues

Starkku committed:
Fix some small issues

Starkku committed:
Allow transparency levels to be interpolated

Starkku committed:
Allow transparency levels to be interpolated

Starkku committed:
Add parsing error handling

Starkku committed:
Add parsing error handling

Starkku committed:
Fix multiflag vector parsing to support redefining keys that have already been added

Starkku committed:
Fix multiflag vector parsing to support redefining keys that have already been added

Starkku committed:
Move general keyframe system docs to Miscellanous.md

Starkku committed:
Move general keyframe system docs to Miscellanous.md

Metadorius commented on docs/New-or-Enhanced-Logics.md:
````diff
@@ -476,6 +476,22 @@ In `artmd.ini`:
+```ini
+[SOMEANIM]                          ; AnimationType
+Translucency.Cloaked=               ; integer - only accepted values are 75, 50, 25 and 0.
+Translucent.KeyframeN.Value=        ; integer - only accepted values are 75, 50, 25 and 0.
+Translucent.KeyframeN.Percentage=   ; floating point value, percents or absolute
+Translucent.KeyframeN.Absolute=     ; integer, zero-based frame index
+Translucent.Interpolation=none      ; Interpolation mode (none|linear)
````
```suggestion
Translucent.KeyframeN.Value=        ; integer - only accepted values are 75, 50, 25 and 0.
Translucent                         ; animatable translucency value - only accepted values are 75, 50, 25 and 0.
```

This way we only need to update one place when we update the template. Also need to add a link to animatable template description.

Accepting ideas on better syntax in the code block.

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1734,3 +1757,210 @@ bool Damageable<T>::Save(PhobosStreamWriter& Stm) const
+	{
+		T dataEntry {};
+
+		// we expect %d for array number then %s for the subflag name, so we replace %s with itself (but escaped)
+		_snprintf_s(flagName, sizeof(flagName), pBaseFlag, i, "%s");
```
I am pretty sure I used `%%` somewhere around here? Just in case, it was not a typo, it's just a way to escape the `%` character because we do formatting twice and need it in the second formatting pass.

Starkku replied:
Yeah but it didn't actually work

Metadorius replied:
:isothonk:

That's weird because the docs say it's the correct way to escape... oh well

Starkku replied:
You don't need to escape it in the args, only in the format specifier if you actually want it to include % as a character. Having double % in the args was causing it to print it twice in the buffer so parsing didn't work.

Metadorius commented on CREDITS.md:
```diff
@@ -245,6 +245,7 @@ This page lists all the individual contributions to the project by their author.
   - Customizing effect of level lighting on air units
   - Reimplemented `Airburst` & `Splits` logic with more customization options
   - Buildings considered as destroyable pathfinding obstacles
+  - Animation transparency customization settings
```
Could you add us as authors of animatable template too?

Starkku committed:
Fix interpolation

The macro approach doesn't work here as long as values are passed by reference

Starkku committed:
Fix interpolation

The macro approach doesn't work here as long as values are passed by reference

Starkku committed:
Update credits

Starkku committed:
Update credits

Starkku committed:
Update docs

Starkku committed:
Update docs

Starkku commented:
Updated credits & docs, ready for another review pass and/or further testing if needed, if not then merge.

Starkku committed:
Add link to feature doc in changelog

Starkku committed:
Add link to feature doc in changelog

Metadorius commented on src/Ext/Anim/Hooks.cpp:
```diff
@@ -469,6 +469,90 @@ DEFINE_HOOK(0x423061, AnimClass_DrawIt_Visibility, 0x6)
+		int frames = pType->End;
+
+		// New addition: Keyframeable Translucent stages.
+		if (pTypeExt->Translucent_Keyframes.KeyframeData.size() > 0)
```
I do think we should make a shortcut function to check whether the template values are set

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1807,3 +1830,210 @@ bool Damageable<T>::Save(PhobosStreamWriter& Stm) const
+TValue Animatable<TValue>::Get(double const percentage) const noexcept
+{
+	// This currently assumes the keyframes are ordered and there are no duplicates for same frame/percentage.
+	// Thing is still far from lightweight as searching for the correct items requires going through the vector.
```
I do think making a cache LUT on the fly based on results of calculation would be a smart idea, just like voxel caching is done.

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1807,3 +1830,210 @@ bool Damageable<T>::Save(PhobosStreamWriter& Stm) const
+
+template<typename T, typename... TExtraArgs>
+requires MultiflagReadable<T, TExtraArgs...>
+void __declspec(noinline) MultiflagValueableVector<T, TExtraArgs...>::Read(INI_EX& parser, const char* const pSection, const char* const pBaseFlag, TExtraArgs&... extraArgs)
```
I wonder whether it is a good idea to add a `.ResetList=true/false` subflag (not serialized, just controls the behavior) here for overriding in map files? If not set - list entries replace the old ones one by one, if set - keyframes are re-set and it's read anew.

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1807,3 +1830,210 @@ bool Damageable<T>::Save(PhobosStreamWriter& Stm) const
+	percentageTemp.Read(parser, pSection, flagName);
+
+	_snprintf_s(flagName, sizeof(flagName), pBaseFlag, "Absolute");
+	absoluteTemp.Read(parser, pSection, flagName);
```
We probably should check whether absoluteLength is not 0 and if it is 0 then not read absolute at all

Metadorius reviewed (changes requested):
@Starkku @ZivDero @CrimRecya What do you think of robustness of the design for the use of multiflag vectors with LaserTrailDataEntry, for example? And other multiflag vectors if you can think of any.

LGTM otherwise

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1807,3 +1830,210 @@ bool Damageable<T>::Save(PhobosStreamWriter& Stm) const
+
+	// we expect "BaseFlagName.%s" here
+	_snprintf_s(flagName, sizeof(flagName), pBaseFlag, "Keyframe%d.%s");
+	this->KeyframeData.Read(parser, pSection, flagName, absoluteLength);
```
Also remind me, what does the template return if there's no values? It probably should return the default for T?

Starkku committed:
Throw an error on trying to instantiate generic type interpolation template

Starkku committed:
Throw an error on trying to instantiate generic type interpolation template

Starkku committed:
Only parse absolute value if absolute length > 0

Starkku committed:
Only parse absolute value if absolute length > 0

Starkku committed:
Add helper function for checking if template has values

Starkku committed:
Add helper function for checking if template has values

Starkku committed:
Refactor / improve parsing and fetching values

- Sort keyframe data after parsing
- Use binary search to lookup values
- Use std::map as a caching mechanism to store values for future lookups

Starkku committed:
Refactor / improve parsing and fetching values

- Sort keyframe data after parsing
- Use binary search to lookup values
- Use std::map as a caching mechanism to store values for future lookups

Starkku committed:
Add a option to reset keyframe data before parsing

Starkku committed:
Add a option to reset keyframe data before parsing

Starkku committed:
Add fallback to parse single value from base key if no keyframes are defined

Starkku committed:
Add fallback to parse single value from base key if no keyframes are defined

Starkku committed:
Refactor / improve animation transparency logic

- Change keyframeable transparency to work under Translucency instead of Translucent to be more consistent with the template's function
- Use Animatable template for both regular and cloak Translucency
- Clean up the hook code, particularly in regards to TranslucencyLevel

Starkku committed:
Refactor / improve animation transparency logic

- Change keyframeable transparency to work under Translucency instead of Translucent to be more consistent with the template's function
- Use Animatable template for both regular and cloak Translucency
- Clean up the hook code, particularly in regards to TranslucencyLevel

Starkku commented:
Updated with feedback addressed.

The translucency customization code has been refactored a bit, instead of keyframes affecting `Translucent` it is now tied to `Translucency` to be more consistent with how the template works with animatable values. `Translucency.Cloaked` can also now be keyframed.

Metadorius reviewed (approved):
LGTM

Metadorius commented on src/Utilities/TemplateDef.h:
```diff
@@ -1870,3 +1893,262 @@ bool Damageable<T>::Save(PhobosStreamWriter& Stm) const
+	char flagName[0x40];
+	for (size_t i = 0; ; ++i)
+	{
+		T dataEntry {};
```
rules:
```ini
Some0.Bull=1
Some0.Shit=2
Some1.Bull=3
Some1.Shit=4
```

map:
```ini

Some1.Bull=0
```

expected result:
```ini
Some0.Bull=1
Some0.Shit=2
Some1.Bull=0
Some1.Shit=4
```

I think it currently will result in
```ini
Some0.Bull=1
Some0.Shit=2
Some1.Bull=0
Some1.Shit=<default>
```

which is not expected

I propose to change the vector to not create a new `T dataEntry {}` always but use the existing one if it exists, and to add `.ResetValue` but per each entry; the only thing it would do is force `T dataEntry {}` always.

Also should move the whole vector ResetValues here from Animatable

Starkku committed:
Improve multiple parsing pass behaviour

- Rename `ResetData`  to `ResetValues` and move from `Animatable` to `MultiflagValueable/NullableVector`
- Make `MultiflagValueable/NullableVector` retain existing values on further parsing passes if unchanged unless explicitly reset.
- Clean up repetitive code in  `MultiflagValueable/NullableVector` with helper functions.
- Disallow < 0 values for `KeyFrameData.Percentage `, throw fatal error on parsing if present.
- Allow parsing `KeyFrameData` on further parsing passes if `Percentage` was parsed on previous pass but not on current.
- Fix base flag name getting cut on the Animatable parsing error fatal error message box.
- Place sorted keyframes in separate vector from parsed values, recalculate on every parsing pass to ensure data correctness.

Starkku commented:
Implemented following changes:

- Rename `ResetData`  to `ResetValues` and move from `Animatable` to `MultiflagValueable/NullableVector`
- Make `MultiflagValueable/NullableVector` retain existing values on further parsing passes if unchanged unless explicitly reset.
- Clean up repetitive code in  `MultiflagValueable/NullableVector` with helper functions.
- Disallow < 0 values for `KeyFrameData.Percentage `, throw fatal error on parsing if present.
- Allow parsing `KeyFrameData` on further parsing passes if `Percentage` was parsed on previous pass but not on current.
- Fix base flag name getting cut on the Animatable parsing error fatal error message box.
- Place sorted keyframes in separate vector from parsed values, recalculate on every parsing pass to ensure data correctness.

Note about the sorting change: I did explore other solutions but they basically boiled down to the following and every single one of them had in my opinion more issues than the chosen approach.
1.  Sort once on first call to `Get()`: Clean but adds runtime during game itself and not just to loading screen.
2. Sort once after all parsing passes are done: Identifying the final parsing pass reliably is not easily doable, it can be up to 4 times (rules, non-client gamemode INI, map file, `tmcj4f.ini` if the cheat option is on) but at the minimum it is 1.
3. Sort once manually on some kind of 'finalize' function: Manually keeping track of all `Animatable` instances, adding some mechanism after all parsing has been done etc. would be way too much work and potentially way more annoying for devs to use as well.
4. Remove binary search: A possibility but ironically enough linear search solutions that are not terrible for performance are harder to understand and read.

And in case it wasn't clear why this had to be changed to begin with, sorting the parsed data always after each parsing pass was causing problems with the data if re-parsed (e.g keyframe indices no longer match originals, possibly)

Starkku committed:
Fix an error

Metadorius committed:
abandon passing down the template screwery

Starkku committed:
Fix fallback value behaviour and adjust docs

Metadorius committed:
simplify reading existing elements
